client/allocrunner/taskrunner/connect_hook.go

```diff
@@ -0,0 +1,145 @@
+package taskrunner
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	agentconsul "github.com/hashicorp/nomad/command/agent/consul"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+var _ interfaces.TaskPrestartHook = &connectHook{}
+
+// connectHook writes the bootstrap config for the Connect proxy sidecar
+type connectHook struct {
+	alloc          *structs.Allocation
+	consulHTTPAddr string
+
+	logger log.Logger
+}
+
+func newConnectHook(alloc *structs.Allocation, consulHTTPAddr string, logger log.Logger) *connectHook {
+	h := &connectHook{
+		alloc:          alloc,
+		consulHTTPAddr: consulHTTPAddr,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (connectHook) Name() string {
+	return "connect"
+}
+
+func (h *connectHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	if !req.Task.Kind.IsConnectProxy() {
+		// Not a Connect proxy sidecar
+		resp.Done = true
+		return nil
+	}
+
+	serviceName := req.Task.Kind.Value()
+	if serviceName == "" {
+		return fmt.Errorf("Connect proxy sidecar does not specify service name")
+	}
+
+	tg := h.alloc.Job.LookupTaskGroup(h.alloc.TaskGroup)
+
+	var service *structs.Service
+	for _, s := range tg.Services {
+		if s.Name == serviceName {
+			service = s
+		}
+	}
+
+	if service == nil {
+		return fmt.Errorf("Connect proxy sidecar task exists but no services configured with a sidecar")
+	}
+
+	h.logger.Debug("bootstrapping Connect proxy sidecar", "task", req.Task.Name, "service", serviceName)
+
+	canary := false
+	if h.alloc.DeploymentStatus != nil {
+		canary = h.alloc.DeploymentStatus.Canary
+	}
+
+	//TODO(schmichael) relies on GRPCSocket being created
+	//TODO(schmichael) unnecessasry if the sidecar is running on the host netns
```

Comment:
Not sure we want to leave TODOs sprinkled around for every place we _break_ Connect-outside-a-namespace... Should make it easier to support Connect on hosts in the future, but I'm not even sure we'll do that.

Reply:
I think its better to have them for now. We can remove them if we feel like it before cutting 0.10